Add support for embedding subwindows and dual screen GameViews

RobProductions opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/11091

Thanks to guidance from @Hilderin this PR adds two new features to the Game tab and embedding tools that were introduced in Godot 4.4. The first change extends the embedding functionality to allow embedding of "subwindows" (as they're known in `DisplayServer`) when provided with a unique window name. The user can now select whether to embed the main window or a custom window that matches the name provided via a text box in the GameView:

https://github.com/user-attachments/assets/3abf1bf9-cbf6-422e-9cfa-a80aef3a6843

This allows developers of multi-window games to embed the game window that makes the most sense for their project rather than being stuck with the main window. This is accomplished through modifications to the display server so that a subwindow can be parented much the same as the main window. In order to send the parenting data to the built game, a new `--swid` (subwindow ID) command line parameter was created that matches the `--wid` implementation for the main window. Then, when the `GameView` modifies the instance parameters, it can optionally send the `--swid` with the user provided window name.

As suggested by @Hilderin , I took the next step in the proposal to better utilize subwindow embedding here by adding an option to split the Game tab into two GameViews for better visibility while debugging. This was largely done by splitting out functionality that used to be handled by the `GameView` into the `GameViewPlugin` and having the plugin manage two instances, but all together it works pretty cleanly! A button in the embed options popup lets you split the view and the layout is saved as project metadata, with the default setting being the regular single view that people are familiar with:

https://github.com/user-attachments/assets/462f88d1-00c5-409b-9334-e6716db84516

Since this PR was already looking like a lot to review, one step I left for a future PR was getting the secondary view's game debugger working properly. This proved challenging to do immediately because we would either need to make the RuntimeSelect stuff not a singleton or we'd need to pass window data through every message sent via the debugger, but I'm certain that one of these solutions will be easier to work out when this initial PR is wrapped up. Besides that, the regular debugger controls for the main GameView are still fully functional:

https://github.com/user-attachments/assets/d6c88390-028c-46e4-86bf-3a5d54ccdbd0

**Note:** This PR only implements subwindow embedding for the Windows display server. Sadly I don't have a Linux machine to test on but Hilderin said it was fine to leave it like this so that a follow up PR can tackle the Linux/Android support which I believe shouldn't be too tough to match from Windows.

Thank you to everyone who offered ideas about the future of embedding in https://github.com/godotengine/godot/pull/99010 and I hope this new functionality proves to be helpful for developers looking for more debugging options! Let me know if there's anything I've missed and I'll be happy to make adjustments :)

RobProductions committed:
Add embedded window selector & dual game view

This commit allows the `GameView` to select either the "Main Window" (default view of the running game) or a subwindow to become embedded when running the project. This information is communicated to the built game via a new `--swid` (subwindow ID) command line parameter which matches the `--wid` implementation created for the original embedding feature. The subwindow is chosen by matching the name of the window to a user provided name which is given in a text box in the `GameView`.

This is implemented only for Windows and a follow up PR should be able to match the functionality for Linux easily.

Additionally, to utilize this new behavior, there is now the option to enable a second `GameView` that sits above the main one. The second view can embed a different window and has its own embedding controls to improve user debugging. For now, in order to keep the PR smaller, the secondary `GameView` does not utilize its own SceneDebugger, so only the main view can use the runtime selection options. A future PR can open up more functionality for parity with the main view.

@jaydensipe also contributed an alternate layout option for the split game view in this commit.

Co-Authored-By: Jayden Sipe <47132280+jaydensipe@users.noreply.github.com>

RobProductions commented:
Rebased to handle the new mute audio button. Like the other debugger controls, it works the same as before on the main game view and is hidden on the secondary view for now to prevent confusion :)

RobProductions commented:
Rebased to handle the feature profile changes. Thanks to the update, there are less `#ifdef` checks now in the new logic

YeldhamDev reviewed (changes requested):
On my end (Fedora 41 GNU/Linux), this breaks a few things:
- Clicking the "Game" tab doesn't focus the window when floating is enabled, instead it switches to it, in a blank state.
- Running with "Custom Window" makes so that no window is embedded at all.

RobProductions commented:
>  * Clicking the "Game" tab doesn't focus the window when floating is enabled, instead it switches to it, in a blank state.

Thanks for trying it out, you're right that this became broken somehow. It was working correctly when I first created the PR but somewhere along rebasing it got messed up, so it should be an easy fix!

>  * Running with "Custom Window" makes so that no window is embedded at all.

You mean setting the dropdown to "Custom Window" and leaving the title blank? On Windows it results in the Main Window getting embedded:

![image](https://github.com/user-attachments/assets/50fdc49e-24a9-4bbf-af10-7d16112edabe)

Note that I haven't implemented subwindow embedding on Linux, but this "empty title" case should give the same result because currently Linux should always embed the main window (only the displayserver signatures were changed). If this is what's happening, I might need to look into this further and get some help testing on Linux.

However, if you put some string into the custom window title field then this would be the expected result on Linux until someone (probably @Hilderin ) matches the Windows implementation and verifies it in a future PR :)

YeldhamDev commented:
I see. Then this leaves only the empty "Game" tab then.

RobProductions commented:
@YeldhamDev Understood. I've updated the code to fix the tab switching issue now, and introduced a new Display Server feature called `FEATURE_SUBWINDOW_EMBEDDING` to handle the difference on Linux and other platforms that haven't caught up yet. The easiest path was to just hide the window selector dropdown and enforce GameView to always embed the main view, hopefully that works as a stopgap for now. Note for whoever implements the Linux side of things: you'll need to first enable the feature in the X11 display server code in order to test subwindow embedding properly.

Feel free to give a try and let me know if this looks better!

YeldhamDev commented:
You should probably hide the dual view option as well.

RobProductions commented:
Apologies for the delay, I've rebased the branch now and also updated the code to disable the dual screen options if `FEATURE_SUBWINDOW_EMBEDDING` is false. So now there should basically be no change on any platforms other Windows until we get update the display server. @YeldhamDev feel free to check it out and let me know if this looks good :)

YeldhamDev reviewed (approved):
The issues I pointed out were fixed. Still, it's better if someone on a platform where this feature is implemented would give a proper review.

jaydensipe commented:
Hey, love the idea of this PR and have been messing around with it, and it works well. A couple of things I wanted to open for discussion.

1. The idea of having a horizontal layout to the windows. I went ahead and implemented this in my branch (https://github.com/jaydensipe/godot/commit/c27c7a22c655e92542c7f2f19c69461e33ba1628) if you wanted to use the code and include it in this commit or just use it as a reference (or as a separate PR).

https://github.com/user-attachments/assets/ecfce118-c289-4639-bc30-bccee2dccc09

2. I think I saw you mentions this somewhere but I can't remember, but having another option to `Main Window` and `Custom Window` being `Debug Window` that will automatically create a Window and add a Camera to the main scene. I think having it automatically do it for the user will make this feature easier for people to use.
3. This one is probably beyond just a suggestion, but in addition to the `Debug Window`, I wonder if we could have a `Scene View` option, which would show the Scene viewport, like Unity. No idea if this is actually technically possible at this point, but something to wonder about. Something like this (mockup):

![image](https://github.com/user-attachments/assets/e19e2b94-b05c-4e45-bf07-a4ae8180319a)

I may play around with 2 and 3, but just wanted to post here just incase.

RobProductions commented:
> The idea of having a horizontal layout to the windows. I went ahead and implemented this in my branch ([jaydensipe@51a6331](https://github.com/jaydensipe/godot/commit/51a6331f165ab12a051035adb4123e5b8b3783ca)) if you wanted to use the code and include it in this commit or just use it as a reference (or as a separate PR).

Wow, this is awesome, thank you so much! I think is really useful and had considered it a while ago but tried to keep the scope down for my initial attempt haha. Would be nice to have it in for this PR! If you'd like, I can paste in your changes to this commit and add you as a co-author? That seems the easiest path for me, or if you have another idea/want your own PR that works too.

> I think I saw you mention this somewhere but I can't remember, but having another option to `Main Window` and `Custom Window` being `Debug Window` that will automatically create a Window and add a Camera to the main scene. I think having it automatically do it for the user will make this feature easier for people to use.

Yup it was discussed in the proposal here: https://github.com/godotengine/godot-proposals/issues/11091 The work needed for that is definitely something a future PR should handle, but I've got ideas on how to tackle it after this. In short, we could send a debugger signal to create a Window Node with a custom name and reuse the embedded subwindow functionality to set the intended window name to that window. Next step after that would be activating the debugger for the top view so that either can be used for debugging, but that definitely requires a new PR. We'd either need to make the game debugger code instantiable (so there can be 2 copies of it) or send the window name along with each signal so that the debugger can act depending on which view the user interacted with; either option should work from what I can tell.

> This one is probably beyond just a suggestion, but in addition to the `Debug Window`, I wonder if we could have a `Scene View` option, which would show the Scene viewport, like Unity. No idea if this is actually technically possible at this point, but something to wonder about. Something like this (mockup):

I could see that being useful, though users have to make sure they don't confuse the Scene View with the debug view - I'm thinking eventually (hopefully) we will have a lot of the same tools for runtime editing like the transform gizmos, so it will be important to distinguish them for beginners who may expect the Scene View to also work in runtime like in Unity. Side note: if it was up to me, I'd prefer an approach different than window embedding in the first place to get the editor tools at runtime, but considering this is what we've got, I'll work with it lol. 

Anyways, I think the long term solution to this would be something like https://github.com/godotengine/godot-proposals/issues/7233 where any viewport can be moved to any position, including the game tab views and 3D plugin so that users can make any layout they want. This is a lot of work but honestly it would probably be better to look into this than to try and figure out all the combinations of side-by-side windows that users might want. That's just my personal take though and if it's too far away, we can work with these alternatives for now

jaydensipe commented:
> If you'd like, I can paste in your changes to this commit and add you as a co-author?
> 

Sure, that would be awesome!

> Yup it was discussed in the proposal here: [godotengine/godot-proposals#11091](https://github.com/godotengine/godot-proposals/issues/11091) The work needed for that is definitely something a future PR should handle, but I've got ideas on how to tackle it after this. In short, we could send a debugger signal to create a Window Node with a custom name and reuse the embedded subwindow functionality to set the intended window name to that window. Next step after that would be activating the debugger for the top view so that either can be used for debugging, but that definitely requires a new PR. We'd either need to make the game debugger code instantiable (so there can be 2 copies of it) or send the window name along with each signal so that the debugger can act depending on which view the user interacted with; either option should work from what I can tell.

> I could see that being useful, though users have to make sure they don't confuse the Scene View with the debug view - I'm thinking eventually (hopefully) we will have a lot of the same tools for runtime editing like the transform gizmos, so it will be important to distinguish them for beginners who may expect the Scene View to also work in runtime like in Unity. Side note: if it was up to me, I'd prefer an approach different than window embedding in the first place to get the editor tools at runtime, but considering this is what we've got, I'll work with it lol.
> 
> Anyways, I think the long term solution to this would be something like [godotengine/godot-proposals#7233](https://github.com/godotengine/godot-proposals/issues/7233) where any viewport can be moved to any position, including the game tab views and 3D plugin so that users can make any layout they want. This is a lot of work but honestly it would probably be better to look into this than to try and figure out all the combinations of side-by-side windows that users might want. That's just my personal take though and if it's too far away, we can work with these alternatives for now

Alright, sweet. Sounds like we share the same hope that we get those debugging gizmos, etc. This PR definitely takes us one step closer. Being able to have the more modular editor would be sick, but I bet you this horizontal/vertical two window is what almost all users will end up using anyways (of course that could be completely false lol).

RobProductions commented:
I've now added in the horizontal split feature from @jaydensipe and it works great! Thanks again for contributing 😄 

https://github.com/user-attachments/assets/4d92495c-5dec-4556-9ee6-92b2e6e307f1

One tiny caveat is the limited space on the top bar which locks the minimum size of the right view while in the horizontal mode, and I can only imagine that this will get worse as we add more debugging buttons. This seems to be a general problem with the Godot editor viewports (I've noticed it in other editor windows too) and I'm not exactly sure what the solution is, but that's a future concern we can reevaluate later.

jaydensipe commented:
> One tiny caveat is the limited space on the top bar which locks the minimum size of the right view while in the horizontal mode, and I can only imagine that this will get worse as we add more debugging buttons. This seems to be a general problem with the Godot editor viewports (I've noticed it in other editor windows too) and I'm not exactly sure what the solution is, but that's a future concern we can reevaluate later.

Good point. I wanna say later we could maybe chop off the "multi selection" tool, since it's already a shortcut on the "selection" tool, but not sure.

AThousandShips commented on editor/plugins/game_view_plugin.h:
```diff
@@ -36,11 +36,16 @@
 #include "scene/gui/box_container.h"
+#include "scene/gui/line_edit.h"
+#include "scene/gui/option_button.h"
+#include "scene/gui/split_container.h"
```
Forward declare these

AThousandShips commented on editor/plugins/game_view_plugin.cpp:
```diff
@@ -551,36 +563,40 @@ void GameView::_update_ui() {
 				state_label->set_text(TTR("Game running not embedded."));
 			} else if (embed_on_play) {
-				state_label->set_text(TTR("Press play to start the game."));
+				if (get_embedded_window_title().is_empty() && main_window_game_view != nullptr && main_window_game_view != this) {
```
```suggestion
				if (main_window_game_view != nullptr && main_window_game_view != this && get_embedded_window_title().is_empty()) {
```

RobProductions commented:
@AThousandShips thanks for the review! I've updated the branch with all of your requested changes and rebased it, so feel free to let me know if this is good and there's anything else I should adjust

RobProductions commented:
It took some trial and error with the CI but I've rebased the branch to account for the MacOS embedding that was added in https://github.com/godotengine/godot/pull/105884 . Like the other platforms, since the mac display server does not set the `FEATURE_SUBWINDOW_EMBEDDING` yet, the window selector & dual screen option won't show up there. Hopefully this gets another review and can be merged soon since it's a becoming a larger effort to keep it free of merge conflicts 😅

RobProductions commented:
Rebased again!

Robert-K commented:
Does this PR also enable embedding two run instances of a game within the Godot Editor?
If not, I'd like to kindly request that.

### Here's my usecase:
I'm building a mobile game with multiplayer ([using the fantastic EOSG plugin](https://godotengine.org/asset-library/asset/2453)) and frequently have to debug/test with two run instances.
The first instance/session gets embedded in the Godot Editor window, but the second one remains a floating window:
<img width="2363" height="1733" alt="image" src="https://github.com/user-attachments/assets/678eba46-d961-48c3-a756-8d7e13c29480" />
Doing anything in the editor will move it to the front, hiding the floating window. This results in obscene amounts Alt+Tabbing.

It would be amazing if the game view provided options for displaying multiple run instances/sessions, with the horizontal/vertical splits from this PR. To be really feature complete, this would probably need to be able to handle more than two sessions as well...

I wouldn't mind iterating on the PR, if that's welcome.

RobProductions commented:
Just rebased the branch and updated some UI to make it play nicely with the new theme:

<img width="1550" height="1012" alt="Screenshot 2025-11-21 133530" src="https://github.com/user-attachments/assets/3160dc82-8823-4227-bd7f-7bad402fa59a" />

The FlowContainer seems to have solved our earlier concerns about the GameView width being restricted by button count, and on top of that the sizing buttons being in a dropdown has reduced the amount of buttons to worry about, so it looks pretty good to me :) Let me know if there's anything more I should adjust!

ZycaR commented:
<img width="273" height="275" alt="image" src="https://github.com/user-attachments/assets/0e40be29-443b-4fbd-89d7-f0d8b4213355" />

akien-mga commented:
CC @bruvzg to review the general DisplayServer and platform-specific parts, and @deralmas to check that it works on Wayland.

deralmas commented:
> Not sure how easy it is to replicate the Windows code but it's a small change so hopefully not too bad!

We have a completely different architecture over at the Wayland side but hopefully it won't require too much work ;)

> Correct, it's not exposed to games. When the editor runs a debugging build of the game with intent to embed a subwindow, it passes the --swid flag in the command line params which is picked up by Core, which then stores a list of subwindows to embed. When a subwindow is created, it checks if the name matches something in the list and if so adds in the p_parent when calling the display server code. This is basically the same as how it works with the regular --wid flag used by main window embedding. Theoretically, a user could create a window at runtime via code that has the same name as one passed from the --swid flag and get that embedded (so long as the editor waits long enough to check for it), but that could be a useful feature for some projects

Thank you for the explanation, now I understand it a lot better :D

Interestingly, since on Wayland we intercept window creation messages, the embedding is done by the editor, you could say. That means that we should either add some control logic to the client, which right now is basically unmodified, or perhaps signal to the `DisplayServer` about the list of names to automatically embed.

Also, do I understand correctly that the "to-embed" list is pre-determined at startup and that windows are embedded only on creation? Right now I've always assumed that things could switch up on the fly, but if that were not the case it would make things a lot easier.

deralmas commented:
> Yup! The Core checks for window names based on --swid at startup and that's the only system affecting the list for now at least. Windows are only embedded on creation because the relevant display server for the platform sets the "window parent" at that time.

Ohhh that's great news, then it will take a lot less work hopefully to port it to the Wayland embedder :D

> So if I'm understanding right, Wayland would just have to make sure that subwindows receive a parent when created, though I would've thought this would be set up for you already when this line happens:

That's the thing, we can't control "foreign" windows, _normally_ at least. That's part of Wayland's design, and it allows for weird new features and whatnot.

The details are _long_ (if you're curious, you can check #107435), but in short we solve this by intercepting the embedded game's requests and mess with its windows without it even knowing. This works because we can change the game's environment variables before spawning it.

That's why I say that the editor does the embedding, we have our own blend of magic for embedding, completely different from the Windows/X11 backend.

That said, yea, I think it's safe to say that with this approach it's more a matter of elbow grease than anything, and it shouldn't take that much time either :D

bruvzg commented on scene/main/window.cpp:
```diff
@@ -665,10 +665,15 @@ void Window::_make_window() {
 
-	window_id = DisplayServer::get_singleton()->create_sub_window(DisplayServer::WindowMode(mode), vsync_mode, f, window_rect, is_in_edited_scene_root() ? false : exclusive, transient_parent ? transient_parent->window_id : DisplayServer::INVALID_WINDOW_ID);
+	// Add parent window if --swid told us to embed a window with this title.
+	int64_t parent_handle = Engine::get_singleton()->get_embedded_subwindow(get_title());
```
I would use something more unique, like node path (`get_path()`) to identify windows (title still can be shown in the user facing UI).

bruvzg replied:
I guess it is fine for the most cases.

bruvzg commented on platform/windows/display_server_windows.cpp:
```diff
@@ -3117,6 +3123,12 @@ static BOOL CALLBACK _enum_proc_find_window_from_process_id_callback(HWND hWnd,
 		}
 
+		wchar_t this_window_title[256];
+		GetWindowTextW(hWnd, this_window_title, sizeof(this_window_title) / sizeof(this_window_title[0]));
```
Same here, instead of using a title which is not unique, it's probably better to use `SetWindowLongW()` and store the unique ID in extra window data (which can be the same ID used as `p_parent/p_embedding_id` in `create_window`).

bruvzg reviewed (commented):
My main concern about `DisaplyServer` API is using window title as ID, which is not unique and can change.

RobProductions commented:
It has been nearly 6 months since @bruvzg made the review comments above and every attempt I made to reach out was met with silence so I will mark the conversation is resolved so that no one is confused about the state of the PR. It's now been over a year since I submitted the branch and I just can't dedicate any more time to it until someone actually reviews the code and confirms that it is a candidate for merging. The conflicts are piling up again and I never expected that it would take this long, yet I'm still getting asked about the state of it by regular users who are excited about this feature.

At this point any communication on the timeline here would be welcome. The last we heard from @akien-mga was that @bruvzg is supposed to review the DisplayServer and @deralmas is to review the Wayland side. The latter review was completed and the former review is stuck in this state of silence, so it is probably time for a new review unless it's just... ready now? There was a new request for review 2 months ago but nothing has changed still. I don't want to keep bugging people to check it out and review, but like every month we wait is more merge conflicts that need resolving, so... 😅 I appreciate any help that can be provided here!

bruvzg commented:
Sorry, not sure why I missed the previous pings.

bruvzg reviewed (commented):
Windows `DisplayServer` code changes look fine. Needs rebase. I'll test it on this weekend.

bruvzg commented:
With the current embedding implementation, it won't be possible to do exactly the same for macOS, since embedded process can create any native windows.

deralmas commented:
> @deralmas is to review the Wayland side

Not sure what I need to review in this case :sweat_smile:

As I said it should be perfectly doable, especially if the window is embedded always on creation and doesn't switch roles.

You don't need to worry about not having a Wayland implementation yet, that's why I care so much about the feature flag. I'll make a follow-up PR later with all the relevant embedder changes.

RobProductions commented:
> Not sure what I need to review in this case 😅

Yeah nothing more is needed there, I was just summarizing the previous conversation, thanks :)

> With the current embedding implementation, it won't be possible to do exactly the same for macOS, since embedded process can create any native windows.

Not sure I understand, are you saying it only makes "native windows" and creating 2 "native windows" not possible for some reason? Or that native window is an option on Mac which will be incompatible with this? Regardless, it's been flagged off on Mac for now via the `FEATURE_SUBWINDOW_EMBEDDING` explained above, so hopefully that's okay for now and once this is all set in stone we could see what's possible there. I appreciate the help!

bruvzg commented:
> Not sure I understand, are you saying it only makes "native windows" and creating 2 "native windows" not possible for some reason?

Currently, macOS has two fully separate display server implementations:
- one is a normal GUI app that can create native windows.
- another one (this one used for embedded game) is non-gui app that can't create any windows (or directly interact with the user) at all, and use shared framebuffer to send rendered content to the editor / network sockets to receive input from editor.

So it can be easily updated to embed multiple windows, but can't have embedded and non-embedded window, to support this we'll need to merge both implementations, which is non-trivial.

Embedding in the way Windows do it (attaching one process native window to another process window) is not possible due to OS restrictions.

> so hopefully that's okay for now and once this is all set in stone we could see what's possible there.

Yes, it's OK to have it implement for Windows only for now.

bruvzg commented:
Tried it on Windows (have not rebase due to a lot of conflicts, so tested as is, and it might be caused by some unrelated issue that was fixed in master), and embedded window seems to jump to random places when host window is resize or moved.

<img width="1892" height="1033" alt="Screenshot 2026-04-12 115510" src="https://github.com/user-attachments/assets/0f604563-0e27-4332-a511-086667b37fce" />

RobProductions commented:
> Tried it on Windows (have not rebase due to a lot of conflicts, so tested as is, and it might be caused by some unrelated issue that was fixed in master), and embedded window seems to jump to random places when host window is resize or moved.
> <img alt="Screenshot 2026-04-12 115510" width="1892" height="1033" src="https://private-user-images.githubusercontent.com/7645683/577036869-0f604563-0e27-4332-a511-086667b37fce.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzYwMDY1MTEsIm5iZiI6MTc3NjAwNjIxMSwicGF0aCI6Ii83NjQ1NjgzLzU3NzAzNjg2OS0wZjYwNDU2My0wZTI3LTQzMzItYTUxMS0wODY2NjdiMzdmY2UucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDQxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjA0MTJUMTUwMzMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZWY4YTE3NmU4OTkyZDAyNjQ4MmY2MzczMWRhODk3ZmNmYTZiM2U3NjkyNDQ2MmFhMjA1YTVjNGMzYjgwZTZlMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmcmVzcG9uc2UtY29udGVudC10eXBlPWltYWdlJTJGcG5nIn0.W1TBgsGqbnoz8Fkysw96ITON_M5qc5S6JJLb-1ADrYQ">

@bruvzg I can't seem to reproduce that on Windows; I get some flickering and slight position offsets due to the embedded window lagging behind the editor but not what you're seeing:

https://github.com/user-attachments/assets/bb2ac3ec-dfdc-4004-b5a1-264e4c323a62

I will try rebasing to see if it gets resolved. If not though, does the problem correct itself after the resize is complete or does it get stuck like that? If it doesn't get stuck that might be an acceptable artifact for now since the embedded window already has issues keeping up with the parent window by default (unless there was some recent update improving it that I missed)

EDIT: Sorry, I just can't handle the rebase right now, there's too many changes piled up and trying to preserve the improvements that were made in the past 4 months is very complicated when I'm not fully focused on it. I tried using Visual Studio's merge tool to help but it wouldn't let me edit other parts of the branch alongside the conflicts and there's seemingly no way to apply conflict resolutions halfway through the process so I just lost that work and will have to redo it at some point. I will see if I have the motivation/time for it another day, but in the meantime I could really use some help at this point if anyone is able :')
